src/mpid/ch4/shm/ipc/src/ipc_control.c

```diff
@@ -0,0 +1,135 @@
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_IPC_SEND_CONTIG_LMT_FIN_CB);
+
+    IPC_TRACE("send_contig_lmt_fin_cb: complete sreq %p\n", sreq);
+
+    MPIR_Datatype_release_if_not_builtin(MPIDIG_REQUEST(sreq, datatype));
+    MPID_Request_complete(sreq);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_IPC_SEND_CONTIG_LMT_FIN_CB);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPIDI_IPC_send_contig_lmt_rts_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
+{
+    int mpi_errno = MPI_SUCCESS;
+    MPIDI_SHM_ctrl_ipc_send_contig_lmt_rts_t *slmt_rts_hdr = &ctrl_hdr->ipc_contig_slmt_rts;
+    MPIR_Request *rreq = NULL;
+    MPIR_Comm *root_comm;
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_IPC_SEND_CONTIG_LMT_RTS_CB);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_IPC_SEND_CONTIG_LMT_RTS_CB);
+
+    IPC_TRACE("send_contig_lmt_rts_cb: received data_sz 0x%lx, sreq_ptr 0x%lx, "
+              "src_lrank %d, match info[src_rank %d, tag %d, context_id 0x%x]\n",
+              slmt_rts_hdr->data_sz, slmt_rts_hdr->sreq_ptr,
+              slmt_rts_hdr->src_lrank, slmt_rts_hdr->src_rank, slmt_rts_hdr->tag,
+              slmt_rts_hdr->context_id);
+
+    /* Try to match a posted receive request.
+     * root_comm cannot be NULL if a posted receive request exists, because
+     * we increase its refcount at enqueue time. */
+    root_comm = MPIDIG_context_id_to_comm(slmt_rts_hdr->context_id);
+    if (root_comm) {
+        while (TRUE) {
+            rreq = MPIDIG_dequeue_posted(slmt_rts_hdr->src_rank, slmt_rts_hdr->tag,
+                                         slmt_rts_hdr->context_id, 1,
+                                         &MPIDIG_COMM(root_comm, posted_list));
+#ifndef MPIDI_CH4_DIRECT_NETMOD
+            if (rreq) {
+                int is_cancelled;
+                MPIDI_anysrc_try_cancel_partner(rreq, &is_cancelled);
+                if (!is_cancelled) {
+                    MPIR_Comm_release(root_comm);       /* -1 for posted_list */
+                    MPIR_Datatype_release_if_not_builtin(MPIDIG_REQUEST(rreq, datatype));
+                    continue;
+                }
+                /* NOTE: NM partner is freed at MPIDI_anysrc_try_cancel_partner,
+                 * no need to call MPIDI_anysrc_free_partner at completions
+                 */
+            }
+#endif
+            break;
+        }
+    }
+
+    if (rreq) {
+        /* Matching receive was posted */
+        MPIR_Comm_release(root_comm);   /* -1 for posted_list */
+        MPIDIG_REQUEST(rreq, rank) = slmt_rts_hdr->src_rank;
+        MPIDIG_REQUEST(rreq, tag) = slmt_rts_hdr->tag;
+        MPIDIG_REQUEST(rreq, context_id) = slmt_rts_hdr->context_id;
+
+        /* Complete IPC receive */
+        switch (slmt_rts_hdr->ipc_type) {
+#ifdef MPIDI_CH4_SHM_ENABLE_XPMEM
+            case MPIDI_SHM_IPC_TYPE__XPMEM:
+                mpi_errno = MPIDI_IPC_xpmem_handle_lmt_recv(slmt_rts_hdr->mem_handle,
```

Comment:
I don't understand why this function exists.  Once the buffer is exposed, why does XPMEM need to be involved in the data transfer at all?

Reply:
It is used for single v.s. coop protocol selection in XPMEM. For now the coop is XPMEM-custom protocol. Even if we change the coop to be IPC-generic in future, I feel that each submodule still has to make selection by itself because the threshold is different.

Reply:
Besides, we cannot let the sender issue a different packet to separate single v.s. coop, because the decision has to be made by receiver.

Reply:
I already responded to the coop issue above.  That's an incorrect specialization to XPMEM.

The packet type should be managed by the IPC layer.  The code is unnecessarily being specialized to XPMEM, when it can be used for other submodules too.

Reply:
How can we use different chunk size in coop for XPMEM and GPU?

Reply:
That's just an exposed value, similar to `MPIR_CVAR_CH4_XPMEM_LMT_MSG_SIZE`, right?  But the coop protocol itself is unrelated to XPMEM, isn't it?

Reply:
I agree that the coop protocol itself is generic. The challenge is how to do the chunking cleaning, does IPC have to check if the remote is a GPU or CPU buffer then use a different chunk size? The only reason we skip it for now is because of the complexity. I feel that it is better to move it to IPC generic in a separate PR, after merging this PR and the GPU submodule PR.

Reply:
I can sympathize with the desire to get this PR in as soon as possible.  Can the coop protocol be dropped from this PR and added in a separate PR?  As you know, I'm more interested in getting this PR in shape for the GPU additions.  So that'd allow us to do that while keeping the interfaces clean.

Reply:
Sure.